### PR TITLE
[CALCITE-6502] Parser loses position information for Expression3

### PR DESCRIPTION
**Pull request by mihaibudiu**
There was another bug in the parsing of array ITEM operators, which was uncovered by this fix.

### CODE REVIEW

File: core/src/main/codegen/templates/Parser.jj [hunk 1]
```diff
@@ -3980,7 +3968,7 @@ SqlNode Expression3(ExprContext exprContext) :
     {
         if (list1.size() == 1) {
             // interpret as single value or query
-            return list1.get(0);
+            return list1.get(0).clone(list1.getParserPosition());
```
Comment (caicancai) [inline core/src/main/codegen/templates/Parser.jj:3971]: Is clone necessary here?

Comment (mihaibudiu) [reply core/src/main/codegen/templates/Parser.jj:3971]: this is actually the line that fixes the JIRA bug

Comment (caicancai) [reply core/src/main/codegen/templates/Parser.jj:3971]: I understand what you mean

**Review by caicancai — APPROVED** [review]

File: core/src/main/codegen/templates/Parser.jj [hunk 2]
```diff
@@ -3810,8 +3810,12 @@ List<Object> Expression2(ExprContext exprContext) :
                 AddExpression2b(list, ExprContext.ACCEPT_SUB_QUERY)
             |
                 <LBRACKET>
-                itemOp = getItemOp()
-                e = Expression(ExprContext.ACCEPT_SUB_QUERY)
+                ( <OFFSET> { itemOp = SqlLibraryOperators.OFFSET; } <LPAREN> { e = Expression(ExprContext.ACCEPT_SUB_QUERY); } <RPAREN>
```
Comment (NobiGo) [inline core/src/main/codegen/templates/Parser.jj:3813]: Is this change necessary？ Can we add code in the `getItemOp()` method?

Comment (mihaibudiu) [reply core/src/main/codegen/templates/Parser.jj:3813]: I didn't know how to do it that way: this way there is context that a simple item does not need ().

Comment (mihaibudiu) [reply core/src/main/codegen/templates/Parser.jj:3813]: this change is necessary, without this the parser will accept an expression like `a[SAFE_OFFSET 2]`

Comment (caicancai) [reply core/src/main/codegen/templates/Parser.jj:3813]: I tested this change and I think it is more appropriate.

**Review by NobiGo — APPROVED** [review]
